File: ios/Classes/ChromeCastController.swift
//
//  ChromeCastController.swift
//  flutter_cast_video
//
//  Created by Alessio Valentini on 07/08/2020.
//

import Flutter
#if canImport(GoogleCast)
import GoogleCast
#endif

class ChromeCastController: NSObject, FlutterPlatformView {
    
    // MARK: - Internal properties
    
    private let channel: FlutterMethodChannel
#if canImport(GoogleCast)
    private let chromeCastButton: GCKUICastButton
    private let sessionManager = GCKCastContext.sharedInstance().sessionManager
#endif
    // MARK: - Init
    
    init(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        registrar: FlutterPluginRegistrar
    ) {
        self.channel = FlutterMethodChannel(name: "flutter_cast_video/chromeCast_\(viewId)", binaryMessenger: registrar.messenger())
#if canImport(GoogleCast)
        self.chromeCastButton = GCKUICastButton(frame: frame)
#endif
        super.init()
        self.configure(arguments: args)
    }
    
    func view() -> UIView {
#if canImport(GoogleCast)
        return chromeCastButton
#else
        return UIView()
#endif
    }
    
    private func configure(arguments args: Any?) {
        setTint(arguments: args)
        setMethodCallHandler()
    }
    
    // MARK: - Styling
    
    private func setTint(arguments args: Any?) {
        guard
            let args = args as? [String: Any],
            let red = args["red"] as? CGFloat,
            let green = args["green"] as? CGFloat,
            let blue = args["blue"] as? CGFloat,
            let alpha = args["alpha"] as? Int else {
            print("Invalid color")
            return
        }
#if canImport(GoogleCast)
        chromeCastButton.tintColor = UIColor(
            red: red / 255,
            green: green / 255,
            blue: blue / 255,
            alpha: CGFloat(alpha) / 255
        )
#endif
    }
    
    // MARK: - Flutter methods handling
    
    private func setMethodCallHandler() {
        channel.setMethodCallHandler { call, result in
            self.onMethodCall(call: call, result: result)
        }
    }
    
    private func setVolume(args: Any?) {
#if canImport(GoogleCast)
        guard
            let args = args as? [String: Any],
            let volume = args["volume"] as? Float,
            let client = sessionManager.currentCastSession?.remoteMediaClient else {
            return
        }
        client.setStreamVolume(volume);
#endif
    }
    
    private func getPlaybackRate() -> Float? {
#if canImport(GoogleCast)
        return sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.playbackRate
#else
        return 1.0
#endif
    }
    
    private func setPlaybackRate(args: Any?) {
#if canImport(GoogleCast)
        guard
            let args = args as? [String: Any],
            let rate = args["rate"] as? Float,
            let client = sessionManager.currentCastSession?.remoteMediaClient else {
            return
        }
        client.setPlaybackRate(rate).delegate = self
#endif
    }
    
    private func getAudioTrack() -> String? {
#if canImport(GoogleCast)
        guard
            let client = sessionManager.currentCastSession?.remoteMediaClient,
            let tracks = client.mediaStatus?.mediaInformation?.mediaTracks?.filter({(track) -> Bool in track.type == GCKMediaTrackType.audio }),
            let ids = sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.activeTrackIDs
        else {
            return nil
        }
        
        return tracks.first { track in
            ids.contains(track.identifier as NSNumber)
        }?.languageCode
#else
        return nil
#endif
    }
    
    private func setAudioTrack(args: Any?) {
#if canImport(GoogleCast)
        guard
            let args = args as? [String: Any],
            let lang = args["lang"] as? String,
            let client = sessionManager.currentCastSession?.remoteMediaClient,
            let tracks = client.mediaStatus?.mediaInformation?.mediaTracks?.filter({(track) -> Bool in track.type == GCKMediaTrackType.audio }),
            let selectedTrackId = tracks.first(where: {track in track.languageCode == lang})?.identifier as NSNumber?
        else {
            return
        }
        
        let trackIds = tracks.map({ track in
            track.identifier
        })
        var ids = client.mediaStatus?.activeTrackIDs?.filter{id in !trackIds.contains(Int(truncating: id))} ?? []
        ids.append(selectedTrackId)
        client.setActiveTrackIDs(ids).delegate = self
#endif
    }
    
    private func getSubtitleTrack() -> String? {
#if canImport(GoogleCast)
        guard
            let client = sessionManager.currentCastSession?.remoteMediaClient,
            let tracks = client.mediaStatus?.mediaInformation?.mediaTracks?.filter({(track) -> Bool in track.type == GCKMediaTrackType.text }),
            let ids = sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.activeTrackIDs
        else {
            return ""
        }
        
        return tracks.first { track in
            ids.contains(track.identifier as NSNumber)
        }?.languageCode
#else
        return nil
#endif
    }
    
    private func setSubtitleTrack(args: Any?) {
#if canImport(GoogleCast)
        guard
            let args = args as? [String: Any],
            let lang = args["lang"] as? String,
            let client = sessionManager.currentCastSession?.remoteMediaClient,
            let tracks = client.mediaStatus?.mediaInformation?.mediaTracks?.filter({(track) -> Bool in track.type == GCKMediaTrackType.text })
        else {
            return
        }

        if lang.isEmpty {
            client.setActiveTrackIDs([]).delegate = self
        } else {
            guard let selectedTrackId = tracks.first(where: {track in track.languageCode == lang})?.identifier as NSNumber? else {
                return
            }
            
            let trackIds = tracks.map({ track in
                track.identifier
            })
            var ids = client.mediaStatus?.activeTrackIDs?.filter{id in !trackIds.contains(Int(truncating: id))} ?? []
            ids.append(selectedTrackId)
            client.setActiveTrackIDs(ids).delegate = self

        }
#endif
    }
    
    
    private func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
#if canImport(GoogleCast)
        switch call.method {
        case "chromeCast#wait":
            result(nil)
            break
        case "chromeCast#loadMedia":
            loadMedia(args: call.arguments)
            result(nil)
            break
        case "chromeCast#play":
            play()
            result(nil)
            break
        case "chromeCast#pause":
            pause()
            result(nil)
            break
        case "chromeCast#seek":
            seek(args: call.arguments)
            result(nil)
            break
        case "chromeCast#setVolume":
            setVolume(args: call.arguments)
            result(nil)
            break
        case "chromeCast#stop":
            stop()
            result(nil)
            break
        case "chromeCast#isConnected":
            result(isConnected())
            break
        case "chromeCast#duration":
            result(duration())
            break
        case "chromeCast#isPlaying":
            result(isPlaying())
            break
        case "chromeCast#getMediaInfo":
            result(getMediaInfo())
            break
        case "chromeCast#addSessionListener":
            addSessionListener()
            result(nil)
        case "chromeCast#removeSessionListener":
            removeSessionListener()
            result(nil)
        case "chromeCast#position":
            result(position())
        case "chromeCast#getPlaybackRate":
            result(getPlaybackRate())
        case "chromeCast#setPlaybackRate":
            setPlaybackRate(args: call.arguments)
            result(nil)
        case "chromeCast#getAudioTrack":
            result(getAudioTrack())
        case "chromeCast#setAudioTrack":
            setAudioTrack(args: call.arguments)
            result(nil)
        case "chromeCast#getSubtitleTrack":
            result(getSubtitleTrack())
        case "chromeCast#setSubtitleTrack":
            setSubtitleTrack(args: call.arguments)
            result(nil)
        case "chromeCast#performClick":
            chromeCastButton.sendActions(for: .touchUpInside)
            result(nil)
        default:
            result(nil)
            break
        }
#else
        result(nil)
#endif
    }
    
    private func loadMedia(args: Any?) {
#if canImport(GoogleCast)
        guard
            let args = args as? [String: Any],
            let url = args["url"] as? String,
            let mediaUrl = URL(string: url) else {
            print("Invalid URL")
            return
        }
        
        let _title = args["title"] as? String
        let _subtitle = args["subtitle"] as? String
        let _image = args["image"] as? String
        let live = args["live"] as? Bool
        let contentType = args["contentType"] as? String ?? "video/mp4"
        let hlsVideoSegmentFormat = args["hlsVideoSegmentFormat"] as? Int ?? 0
        let customData = args["customData"] as? [AnyHashable: AnyHashable]
        
        let movieMetadata = GCKMediaMetadata()
        
        if let title = _title {
            movieMetadata.setString(title, forKey: kGCKMetadataKeyTitle)
        }
        if let subtitle = _subtitle {
            movieMetadata.setString(subtitle, forKey: kGCKMetadataKeySubtitle)
        }
        if let image = _image {
            if let imageUrl = URL(string: image){
                movieMetadata.addImage(GCKImage(url: imageUrl, width: 480, height: 360))
            }
        }
        
        let mediaInfoBuilder = GCKMediaInformationBuilder.init(contentURL: mediaUrl)
        
        mediaInfoBuilder.metadata = movieMetadata

        mediaInfoBuilder.streamType = .buffered
        // 
        mediaInfoBuilder.contentType = contentType
        //
        mediaInfoBuilder.hlsVideoSegmentFormat = GCKHLSVideoSegmentFormat(rawValue: hlsVideoSegmentFormat) ?? .FMP4
        // 
        mediaInfoBuilder.customData = customData

        // 
        if let islive = live {
            if islive {
                mediaInfoBuilder.streamType = .live
            }
        }
        
        let mediaInformation = mediaInfoBuilder.build()
        
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.loadMedia(mediaInformation) {
            request.delegate = self
        }
#endif
    }
    
    private func play() {
#if canImport(GoogleCast)
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.play() {
            request.delegate = self
        }
#endif
    }
    
    private func pause() {
#if canImport(GoogleCast)
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.pause() {
            request.delegate = self
        }
#endif
    }
    
    private func seek(args: Any?) {
#if canImport(GoogleCast)
        guard
            let args = args as? [String: Any],
            let relative = args["relative"] as? Bool,
            let interval = args["interval"] as? Double else {
            return
        }
        let seekOptions = GCKMediaSeekOptions()
        seekOptions.relative = relative
        seekOptions.interval = interval
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.seek(with: seekOptions) {
            request.delegate = self
        }
#endif
    }
    
    private func stop() {
#if canImport(GoogleCast)
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.stop() {
            request.delegate = self
        }
#endif
    }
    
    private func getMediaInfo() -> [String: AnyHashable?]? {
#if canImport(GoogleCast)
        return  mediaInfoToMap(_mediaInfo: sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.mediaInformation)
#else
        return nil
#endif
    }
    
#if canImport(GoogleCast)
    private func mediaInfoToMap(_mediaInfo: GCKMediaInformation?) -> [String: AnyHashable?]? {
        var info = [String: AnyHashable?]()
        if let mediaInfo = _mediaInfo {
            info["id"] = mediaInfo.contentID
            if let u = mediaInfo.contentURL {
                info["url"] = u.absoluteString
            }
            info["contentType"] = mediaInfo.contentType
            info["audioTracks"] = mediaInfo.mediaTracks?.filter({(track) -> Bool in
                track.type == GCKMediaTrackType.audio
            }).map({(track) -> String in
                track.languageCode ?? "unknown"
            })
            info["subtitleTracks"] = Array(Set(mediaInfo.mediaTracks?.filter({(track) -> Bool in
                track.type == GCKMediaTrackType.text
            }).map({(track) -> String in
                track.languageCode ?? "unknown"
            }) ?? []))
            info["customData"] = mediaInfo.customData as? [String: AnyHashable?]
            if let meta = mediaInfo.metadata {
                
                info["title"] =  meta.string(forKey: kGCKMetadataKeyTitle)
                info["subtitle"] =  meta.string(forKey: kGCKMetadataKeySubtitle)
                let imgs = meta.images()
                if (imgs.count > 0){
                    if let img = imgs[0] as? GCKImage {
                        info["image"] = img.url.absoluteString
                    }
                    
                }
                
                
                
            }
        }
        return info;
    }
#endif
    
    private func isConnected() -> Bool {
#if canImport(GoogleCast)
        return sessionManager.currentCastSession?.remoteMediaClient?.connected ?? false
#else
        return false
#endif
    }
    
    private func isPlaying() -> Bool {
#if canImport(GoogleCast)
        return sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.playerState == GCKMediaPlayerState.playing
#else
        return false
#endif
    }
    
    private func addSessionListener() {
#if canImport(GoogleCast)
        sessionManager.add(self)
        if(isConnected()) {
            sessionManager.currentCastSession?.remoteMediaClient?.add(self)
            channel.invokeMethod("chromeCast#didStartSession", arguments: nil)
        }
#endif
    }
    
    private func removeSessionListener() {
#if canImport(GoogleCast)
        sessionManager.remove(self)
#endif
    }
    
    private func position() -> Int {
#if canImport(GoogleCast)
        return Int(sessionManager.currentCastSession?.remoteMediaClient?.approximateStreamPosition() ?? 0) * 1000
#else
        return 0
#endif
    }
    
    private func duration() -> Int {
#if canImport(GoogleCast)
        return Int(sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.mediaInformation?.streamDuration ?? 0) * 1000
#else
        return 0
#endif
    }
    
}

#if canImport(GoogleCast)
// MARK: - GCKSessionManagerListener
extension ChromeCastController: GCKSessionManagerListener {
    func sessionManager(_ sessionManager: GCKSessionManager, didStart session: GCKSession) {
        session.remoteMediaClient?.add(self)
        channel.invokeMethod("chromeCast#didStartSession", arguments: nil)
    }
    
    func sessionManager(_ sessionManager: GCKSessionManager, didEnd session: GCKSession, withError error: Error?) {
        session.remoteMediaClient?.remove(self)
        channel.invokeMethod("chromeCast#didEndSession", arguments: nil)
    }
}

// MARK: - GCKRequestDelegate

extension ChromeCastController: GCKRequestDelegate {
    func requestDidComplete(_ request: GCKRequest) {
        channel.invokeMethod("chromeCast#requestDidComplete", arguments: nil)
    }
    
    func request(_ request: GCKRequest, didFailWithError error: GCKError) {
        channel.invokeMethod("chromeCast#requestDidFail", arguments: ["error" : error.localizedDescription])
    }
}

// MARK: - GCKRemoteMediaClientListener
extension ChromeCastController : GCKRemoteMediaClientListener {
    func remoteMediaClient(_ client: GCKRemoteMediaClient, didUpdate mediaStatus: GCKMediaStatus?) {
        let playerStatus: GCKMediaPlayerState = mediaStatus?.playerState ?? GCKMediaPlayerState.unknown
        var retCode = 4;
        if (playerStatus == GCKMediaPlayerState.playing) {
            retCode = 1
        } else if (playerStatus == GCKMediaPlayerState.buffering) {
            retCode = 0
        } else if (playerStatus == GCKMediaPlayerState.idle && mediaStatus?.idleReason == GCKMediaPlayerIdleReason.finished) {
            retCode = 2
        }else if (playerStatus == GCKMediaPlayerState.paused) {
            retCode = 3;
        }
        channel.invokeMethod("chromeCast#didPlayerStatusUpdated", arguments: retCode)
    }
}
#endif

File: android/src/main/kotlin/it/aesys/flutter_cast_video/ChromeCastController.kt
package it.aesys.flutter_cast_video

import android.content.Context
import android.net.Uri
import android.view.ContextThemeWrapper
import android.view.View
import androidx.mediarouter.app.MediaRouteButton
import com.google.android.gms.cast.*
import com.google.android.gms.cast.framework.*
import com.google.android.gms.cast.framework.media.RemoteMediaClient
import com.google.android.gms.common.api.PendingResult
import com.google.android.gms.common.api.Status
import com.google.android.gms.common.images.WebImage
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import org.json.JSONArray
import org.json.JSONObject
import java.lang.Exception


class ChromeCastController(
	messenger: BinaryMessenger, viewId: Int, context: Context?
) : PlatformView, MethodChannel.MethodCallHandler, SessionManagerListener<Session>,
	PendingResult.StatusListener {
	private val channel = MethodChannel(messenger, "flutter_cast_video/chromeCast_$viewId")
	private val chromeCastButton =
		MediaRouteButton(ContextThemeWrapper(context, R.style.Theme_AppCompat_NoActionBar))
	private var sessionManager : SessionManager? = null

	init {
		CastButtonFactory.setUpMediaRouteButton(context as Context, chromeCastButton)
		chromeCastButton.visibility = View.GONE
		channel.setMethodCallHandler(this)
		try {
			sessionManager = CastContext.getSharedInstance()?.sessionManager
		} catch (ignored: Exception){}
	}

	private fun loadMedia(args: Any?) {
		if (args is Map<*, *>) {
			val url = args["url"] as? String ?: ""
			val title = args["title"] as? String ?: ""
			val subtitle = args["subtitle"] as? String ?: ""
			val imageUrl = args["image"] as? String ?: ""
			val contentType = args["contentType"] as? String ?: "videos/mp4"
			val hlsVideoSegmentFormat = args["hlsVideoSegmentFormat"] as? Int ?: 0
			val customData = (args["customData"] as Map<*, *>?)?.run {
				JSONObject(this)
			} ?: JSONObject()
			
			val liveStream = args["live"] as? Boolean ?: false

			val movieMetadata = MediaMetadata(MediaMetadata.MEDIA_TYPE_MOVIE)

			val streamType =
				if (liveStream) MediaInfo.STREAM_TYPE_LIVE else MediaInfo.STREAM_TYPE_BUFFERED

			movieMetadata.putString(MediaMetadata.KEY_TITLE, title)
			movieMetadata.putString(MediaMetadata.KEY_SUBTITLE, subtitle)
			movieMetadata.addImage(WebImage(Uri.parse(imageUrl)))
			//movieMetadata.addImage(WebImage(Uri.parse(imageUrl)))

			val media = MediaInfo.Builder(url).setStreamType(streamType).setContentType(contentType)
				.setMetadata(movieMetadata).setCustomData(customData).build()
			val options = MediaLoadOptions.Builder().setCustomData(customData).build()
			val request =
				sessionManager?.currentCastSession?.remoteMediaClient?.load(media, options)

			request?.addStatusListener(this)
		}
	}

	private fun play() {
		val request = sessionManager?.currentCastSession?.remoteMediaClient?.play()
		request?.addStatusListener(this)
	}

	private fun pause() {
		val request = sessionManager?.currentCastSession?.remoteMediaClient?.pause()
		request?.addStatusListener(this)
	}

	/*private fun mediaQueue(args: Any?) : List<HashMap<String,String>>{
	   val items : List<HashMap<String,String>> = mutableListOf()
		val client = sessionManager?.currentCastSession?.remoteMediaClient ?: return items
	   val queue = client.getMediaQueue()
	   val status = client.mediaStatus
	   val qlen = queue.getItemCount()

	   var start = 0
	   var batch = 5
	   var page = 1;
	   if (args is Map<*, *>) {
		   batch = (args["batch"] as? Int) ?: batch
		   page = (args["page"] as? Int) ?: page
	   }
		start = page * batch
		var end = (page+1) * batch
		if (start >= qlen){
			return items
		}
		if (end > qlen){
			end = qlen
		}
		batch = end - start

		var ret = queue.fetchMoreItemsRelativeToIndex(start, batch, 0)
		ret.setResultCallback {

		}
	}
	 */
	private fun seek(args: Any?) {
		if (args is Map<*, *>) {
			val relative = (args["relative"] as? Boolean) ?: false
			var interval = args["interval"] as? Double
			interval = interval?.times(1000)
			if (relative) {
				interval = interval?.plus(
					sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.streamPosition
						?: 0
				)
			}
			val request = sessionManager?.currentCastSession?.remoteMediaClient?.seek(
				MediaSeekOptions.Builder().setPosition(interval?.toLong() ?: 0).build()
			)
			request?.addStatusListener(this)
		}
	}

	private fun mediaInfoToMap(mediaInfo: MediaInfo?): HashMap<String, Any?>? {
		return mediaInfo?.let {
			val info = hashMapOf<String, Any?>()
			val id = mediaInfo.contentId
			info["id"] = id
			info["url"] = mediaInfo.contentUrl ?: id
			info["contentType"] = mediaInfo.contentType ?: ""
			info["customData"] = mediaInfo.customData?.toMap()
			info["audioTracks"] =
				mediaInfo.mediaTracks?.filter { track -> track.type == MediaTrack.TYPE_AUDIO }
					?.map { track -> track.language }?.toList()
			info["subtitleTracks"] =
				mediaInfo.mediaTracks?.filter { track -> track.type == MediaTrack.TYPE_TEXT || track.subtype == MediaTrack.SUBTYPE_SUBTITLES }
					?.map { track -> track.language }?.toList()
			mediaInfo.metadata?.apply {
				info["title"] = getString(MediaMetadata.KEY_TITLE) ?: ""
				info["subtitle"] = getString(MediaMetadata.KEY_SUBTITLE) ?: ""
				val imgs = images
				if (imgs.size > 0) {
					info["image"] = imgs[0].url.toString()
				}
			}

			info
		}
	}

	private fun getMediaInfo(): HashMap<String, Any?>? =
		mediaInfoToMap(sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo)


	private fun setVolume(args: Any?) {
		if (args is Map<*, *>) {
			val volume = args["volume"] as? Double
			val request = sessionManager?.currentCastSession?.remoteMediaClient?.setStreamVolume(
				volume ?: 0.0
			)
			request?.addStatusListener(this)
		}
	}

	private fun getVolume() = sessionManager?.currentCastSession?.volume ?: 0.0

	private fun stop() {
		val request = sessionManager?.currentCastSession?.remoteMediaClient?.stop()
		request?.addStatusListener(this)
	}

	private fun isPlaying() =
		sessionManager?.currentCastSession?.remoteMediaClient?.isPlaying ?: false

	private fun isConnected() = sessionManager?.currentCastSession?.isConnected ?: false

	private fun endSession() = sessionManager?.endCurrentSession(true)

	private fun position() =
		sessionManager?.currentCastSession?.remoteMediaClient?.approximateStreamPosition ?: 0

	private fun duration() =
		sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo?.streamDuration ?: 0

	private fun getSubtitleTrack() =
		sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo?.mediaTracks?.filter { track -> track.type == MediaTrack.TYPE_TEXT || track.subtype == MediaTrack.SUBTYPE_SUBTITLES }
			?.let { tracks ->
				sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.activeTrackIds?.toList()
					?.let { ids ->
						tracks.firstOrNull { track -> ids.contains(track.id) }?.language ?: ""
					}

			}

	private fun getAudioTrackLang() =
		sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo?.mediaTracks?.filter { track -> track.type == MediaTrack.TYPE_AUDIO }
			?.let { tracks ->
				sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.activeTrackIds?.toList()
					?.let { ids ->
						tracks.firstOrNull { track -> ids.contains(track.id) }?.language ?: ""
					}

			}

	private fun setSubtitleTrack(args: Any?) = (args as Map<*, *>?)?.apply {
		(this["lang"] as String?)?.let { lang ->
			sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo?.mediaTracks?.filter { track -> track.type == MediaTrack.TYPE_TEXT || track.subtype == MediaTrack.SUBTYPE_SUBTITLES }
				?.also { tracks ->
					if(lang.isEmpty()) {
						sessionManager?.currentCastSession?.remoteMediaClient?.apply {
							setActiveMediaTracks(
								LongArray(0)
							).addStatusListener { status -> onComplete(status) }
						}
					} else if (tracks.isNotEmpty()) {
						tracks.firstOrNull { track -> track.language == lang }?.also { track ->

							(sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.activeTrackIds?.toMutableList()
								?: mutableListOf()).also { ids ->
								ids.removeAll(tracks.map { track -> track.id }.toSet())
								ids.add(track.id)

								sessionManager?.currentCastSession?.remoteMediaClient?.apply {
									setActiveMediaTracks(
										ids.toLongArray()
									).addStatusListener { status -> onComplete(status) }
								}
							}
						}
					}
				}
		}
	}


	private fun setAudioTrack(args: Any?) = (args as Map<*, *>?)?.apply {
		(this["lang"] as String?)?.let { lang ->
			sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo?.mediaTracks?.filter { track -> track.type == MediaTrack.TYPE_AUDIO }
				?.also { tracks ->
					if (tracks.isNotEmpty()) {
						tracks.firstOrNull { track -> track.language == lang }?.also { track ->

							(sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.activeTrackIds?.toMutableList()
								?: mutableListOf()).also { ids ->
								ids.removeAll(tracks.map { track -> track.id }.toSet())
								ids.add(track.id)

								sessionManager?.currentCastSession?.remoteMediaClient?.apply {
									setActiveMediaTracks(
										ids.toLongArray()
									).addStatusListener { status -> onComplete(status) }
								}
							}
						}
					}
				}
		}
	}


	private fun getPlaybackRate() =
		sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.playbackRate

	private fun setPlaybackRate(args: Any?) = (args as Map<*, *>?)?.apply {
		(this["rate"] as Double?)?.apply {
			sessionManager?.currentCastSession?.remoteMediaClient?.setPlaybackRate(this)
				?.addStatusListener { status -> onComplete(status) }
		}
	}


	private fun addSessionListener() {
		sessionManager?.addSessionManagerListener(this)
		if (isConnected()) {
			onSessionStarted(sessionManager!!.currentCastSession!!, "");
		}
	}

	private fun removeSessionListener() {
		sessionManager?.removeSessionManagerListener(this)
		sessionManager?.currentCastSession?.remoteMediaClient?.unregisterCallback(
			mRemoteMediaClientListener
		)
	}

	private val mRemoteMediaClientListener: RemoteMediaClient.Callback =
		object : RemoteMediaClient.Callback() {
			override fun onStatusUpdated() {
				val mediaStatus: MediaStatus? =
					sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus
				val retCode = when (mediaStatus?.playerState ?: MediaStatus.PLAYER_STATE_UNKNOWN) {
					MediaStatus.PLAYER_STATE_BUFFERING -> 0
					MediaStatus.PLAYER_STATE_PLAYING -> 1
					MediaStatus.PLAYER_STATE_IDLE -> 2
					MediaStatus.PLAYER_STATE_PAUSED -> 3
					else -> if (mediaStatus?.idleReason === MediaStatus.IDLE_REASON_FINISHED) 2 else 4
				}
				channel.invokeMethod("chromeCast#didPlayerStatusUpdated", retCode)
			}

			override fun onMediaError(mediaError: MediaError) {
				val errorCode: Int = mediaError.detailedErrorCode ?: 100
				channel.invokeMethod("chromeCast#didPlayerStatusUpdated", errorCode)
			}
		}

	override fun getView() = chromeCastButton

	override fun dispose() {

	}

	// Flutter methods handling

	override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
		when (call.method) {
			"chromeCast#wait" -> result.success(null)
			"chromeCast#loadMedia" -> {
				loadMedia(call.arguments)
				result.success(null)
			}
			"chromeCast#play" -> {
				play()
				result.success(null)
			}
			"chromeCast#pause" -> {
				pause()
				result.success(null)
			}
			"chromeCast#seek" -> {
				seek(call.arguments)
				result.success(null)
			}
			"chromeCast#setVolume" -> {
				setVolume(call.arguments)
				result.success(null)
			}
			"chromeCast#getMediaInfo" -> result.success(getMediaInfo())
			"chromeCast#getVolume" -> result.success(getVolume())
			"chromeCast#stop" -> {
				stop()
				result.success(null)
			}
			"chromeCast#isPlaying" -> result.success(isPlaying())
			"chromeCast#isConnected" -> result.success(isConnected())
			"chromeCast#endSession" -> {
				endSession()
				result.success(null)
			}
			"chromeCast#position" -> result.success(position())
			"chromeCast#duration" -> result.success(duration())
			"chromeCast#addSessionListener" -> {
				addSessionListener()
				result.success(null)
			}
			"chromeCast#removeSessionListener" -> {
				removeSessionListener()
				result.success(null)
			}
			"chromeCast#getPlaybackRate" -> {
				result.success(getPlaybackRate())
			}
			"chromeCast#setPlaybackRate" -> {
				setPlaybackRate(call.arguments)
				result.success(null)
			}
			"chromeCast#getSubtitleTrack" -> {
				result.success(getSubtitleTrack())
			}
			"chromeCast#setSubtitleTrack" -> {
				setSubtitleTrack(call.arguments)
				result.success(null)
			}
			"chromeCast#getAudioTrack" -> {
				result.success(getAudioTrackLang())
			}
			"chromeCast#setAudioTrack" -> {
				setAudioTrack(call.arguments)
				result.success(null)
			}
			"chromeCast#performClick" -> {
				chromeCastButton.performClick()
				result.success(null)
			}
		}
	}

	// SessionManagerListener

	override fun onSessionStarted(p0: Session, p1: String) {
		if (p0 is CastSession) {
			p0.remoteMediaClient?.registerCallback(mRemoteMediaClientListener)
		}
		channel.invokeMethod("chromeCast#didStartSession", null)
	}

	override fun onSessionEnded(p0: Session, p1: Int) {
		channel.invokeMethod("chromeCast#didEndSession", null)
	}

	override fun onSessionResuming(p0: Session, p1: String) {

	}

	override fun onSessionResumed(p0: Session, p1: Boolean) {

	}

	override fun onSessionResumeFailed(p0: Session, p1: Int) {

	}

	override fun onSessionSuspended(p0: Session, p1: Int) {

	}

	override fun onSessionStarting(p0: Session) {

	}

	override fun onSessionEnding(p0: Session) {

	}

	override fun onSessionStartFailed(p0: Session, p1: Int) {

	}

	// PendingResult.StatusListener

	override fun onComplete(status: Status) {
		if (status.isSuccess) {
			channel.invokeMethod("chromeCast#requestDidComplete", null)
		}
	}
}

fun JSONObject.toMap(): Map<String, Any?> = keys().asSequence().associateWith { key ->
	when (val value = this[key]) {
		is JSONArray -> {
			val map =
				(0 until value.length()).associate { index -> Pair(index.toString(), value[index]) }
			JSONObject(map).toMap().values.toList()
		}
		is JSONObject -> value.toMap()
		JSONObject.NULL -> null
		else -> value
	}
}